CAT/model/NCAT.py

```diff
@@ -0,0 +1,601 @@
+            done = False
+        else:
+            done = True
+        self.short[action] = 1
+        t = self.state[1] + [[action, reward, done]]
+        info = {"ACC": acc,
+                "AUC": auc,
+                "rate":rate}
+        self.state[1].append([action, reward, done, info])
+        return self.state, reward, done, info
+
+    def reward(self, action,sid):
+        items = [state[0] for state in self.state[1]] + [action]
+        correct = [self.rates[self.state[0][0]][it] for it in items]
+        self.tsdata.apply_selection(sid, action)
+        loss = self.model.adaptest_update(self.tsdata)
+        acc,auc=self.model.evaluate(self.tsdata)
+        return -loss, acc, auc, correct[-1]
+
+    def precision(self, episode):
+        return sum([i[1] for i in episode])
+
+
+    def recall(self, episode, uid):
+        return sum([i[1] for i in episode]) / len(self.rates[uid])
+
+    def step_policy(self,policy):
+        policy = policy[:self.args.T]
+        rewards = []
+        for action in policy:
+            if action in self.rates[self.state[0][0]]:
+                rewards.append(self.rates[self.state[0][0]][action])
+            else:
+                rewards.append(0)
+        t = [[a,rewards[i],False] for i,a in enumerate(policy)]
+        info = {"precision": self.precision(t),
+                "recall": self.recall(t, self.state[0][0])}
+        self.state[1].extend(t)
+        return self.state,rewards,True,info
+
+    def ndcg(self, episode, uid):
+        if len(self.rates[uid]) > len(episode):
+            return self.dcg_at_k(list(map(lambda x: x[1], episode)),
+                                 len(episode),
+                                 method=1) / self.dcg_at_k(sorted(list(self.rates[uid].values()),reverse=True),
+                                                           len(episode),
+                                                           method=1)
+        else:
+            return self.dcg_at_k(list(map(lambda x: x[1], episode)),
+                                 len(episode),
+                                 method=1) / self.dcg_at_k(
+                list(self.rates[uid].values()) + [0] * (len(episode) - len(self.rates[uid])),
+                len(episode), method=1)
+
+    def dcg_at_k(self, r, k, method=1):
+        r = np.asfarray(r)[:k]
+        if r.size:
+            if method == 0:
+                return r[0] + np.sum(r[1:] / np.log2(np.arange(2, r.size + 1)))
+            elif method == 1:
+                return np.sum(r / np.log2(np.arange(2, r.size + 2)))
+            else:
+                raise ValueError('method must be 0 or 1.')
+
+    def alpha_dcg(self, item_list, k=10, alpha=0.5, *args):
+        items = []
+        G = []
+        for i, item in enumerate(item_list[:k]):
+            items += item
+            G.append(sum(map(lambda x: math.pow(alpha, x - 1), dict(Counter(items)).values())) / math.log(i + 2, 2))
+        return sum(G)
+    
+class NCATModel():
```

Comment:
It is a model for question selection strategy and not a CD model right? Correct me if I am wrong - I have not read the NCAT paper yet. `/model` should contain code for CDMs and `/strategy` contains code for question selection strategy. If NCATModel is not a CDM, it should not be placed in this folder. Modularize your code